smartlearner/optimizers/optimizer.py

```diff
@@ -12,12 +12,22 @@ def __init__(self, loss):
         self._update_rules = []
         self._param_modifiers = []
 
+        self.graph_updates = OrderedDict()
+        self._directions = None
+
     def append_update_rule(self, update_rule):
         self._update_rules.append(update_rule)
 
     def append_param_modifier(self, param_modifier):
         self._param_modifiers.append(param_modifier)
 
+    @property
+    def directions(self):
+        if self._directions is None:
```

Comment:
I agree but I haven't find the right way to make sure this does not happen. Do you have one in mind?

The main reason of this refactoring is to be able to monitor the gradients as well as the directions (and we should have the same for the loss) so we can print them or compute use them to compute some metrics (e.g. check that the norm of the gradients approaches 0).

Right now, when adding such task the entire graph is not created yet, we only have access to the part coming from the model through `model.get_model_output`. The graph is fully created only once we called `trainer.train` and then it is already too late to add tasks to the trainer.

The best would be to disallow graph modification once the `Trainer`object is created. After that, only tasks can be added to the trainer before calling `trainer.train`. Does that make sense?